src/Workspaces/Core/Portable/Workspace/Solution/DocumentInfo.cs

```diff
@@ -88,70 +70,186 @@ public sealed class DocumentInfo
+            Optional<TextLoader> loader = default(Optional<TextLoader>))
         {
-            var newId = id ?? this.Id;
-            var newName = name ?? this.Name;
-            var newFolders = folders ?? this.Folders;
-            var newSourceCodeKind = sourceCodeKind.HasValue ? sourceCodeKind.Value : this.SourceCodeKind;
-            var newLoader = loader.HasValue ? loader.Value : this.TextLoader;
-            var newFilePath = filePath.HasValue ? filePath.Value : this.FilePath;
-
-            if (newId == this.Id &&
-                newName == this.Name &&
-                newFolders == this.Folders &&
-                newSourceCodeKind == this.SourceCodeKind &&
-                newLoader == this.TextLoader &&
-                newFilePath == this.FilePath)
+            var newAttributes = attributes ?? Attributes;
+            var newLoader = loader.HasValue ? loader.Value : TextLoader;
+
+            if (newAttributes == Attributes && newLoader == TextLoader)
             {
                 return this;
             }
 
-            return new DocumentInfo(newId, newName, newFolders, newSourceCodeKind, newLoader, newFilePath, this.IsGenerated);
+            return new DocumentInfo(newAttributes, newLoader);
         }
 
         public DocumentInfo WithId(DocumentId id)
         {
-            return this.With(id: id);
+            return With(attributes: Attributes.With(id: id));
         }
 
         public DocumentInfo WithName(string name)
         {
-            return this.With(name: name);
+            return this.With(attributes: Attributes.With(name: name));
         }
 
         public DocumentInfo WithFolders(IEnumerable<string> folders)
         {
-            return this.With(folders: folders.ToImmutableReadOnlyListOrEmpty());
+            return this.With(attributes: Attributes.With(folders: folders.ToImmutableReadOnlyListOrEmpty()));
         }
 
         public DocumentInfo WithSourceCodeKind(SourceCodeKind kind)
         {
-            return this.With(sourceCodeKind: kind);
+            return this.With(attributes: Attributes.With(sourceCodeKind: kind));
         }
 
         public DocumentInfo WithTextLoader(TextLoader loader)
         {
-            return this.With(loader: loader);
+            return With(loader: loader);
         }
 
         public DocumentInfo WithFilePath(string filePath)
         {
-            return this.With(filePath: filePath);
+            return this.With(attributes: Attributes.With(filePath: filePath));
         }
 
         private string GetDebuggerDisplay()
         {
             return (FilePath == null) ? (nameof(Name) + " = " + Name) : (nameof(FilePath) + " = " + FilePath);
         }
+
+        /// <summary>
+        /// type that contains information regarding this document itself but
+        /// no tree information such as document info
+        /// </summary>
+        internal class DocumentAttributes : IChecksummedObject, IObjectWritable
```

Comment:
Would a struct be better, since this is just raw data.  It will mean we won't have an extra allocation per document.  (though that probably isn't important enough to optimize).